lib/s3.py

```diff
@@ -40,6 +50,37 @@ def is_key_present(url):
     return get_key(url.hostname) is not None
 
 
+def sign_request(url, method='GET', checksum=SHA256_NIL, headers={}, region='any', service='S3'):
+    access_key, secret_key = get_key(url.hostname)
+
+    amzdate = time.strftime('%Y%m%dT%H%M%SZ', time.gmtime())
+
+    headers = headers.copy()  # don't modify the user's copy, or the default
+    headers.update({'host': url.hostname, 'x-amz-content-sha256': checksum, 'x-amz-date': amzdate})
+    headers_str = ''.join(f'{k}:{v}\n' for k, v in sorted(headers.items()))
+    headers_list = ';'.join(sorted(headers))
+
+    credential_scope = f'{amzdate[:8]}/{region}/{service}/aws4_request'
+    signing_key = f'AWS4{secret_key}'.encode('ascii')
+    for item in credential_scope.split('/'):
+        signing_key = hmac.new(signing_key, item.encode('ascii'), hashlib.sha256).digest()
+
+    algorithm = 'AWS4-HMAC-SHA256'
+    canonical_request = f'{method}\n{url.path}\n{url.query}\n{headers_str}\n{headers_list}\n{checksum}'
```

Comment:
Not your "fault" of course, but can you think of any reason why the list of header names appears here? It's already contained in the full headers_str, after all. It's mentioned in [the docs](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html), but not without much rationale.

Reply:
Yes: The headers_str is part of what gets hashed, which means that it arrives at the server completely opaque.  The headers_list is sent in the plain to help the server decide which headers it should hash on its side, to ensure that we get the same result on both ends.

There are some parts of v4 signing which I find obnoxious (like the needlessly-complicated key derivation stuff) but this particular feature of making the list of headers explicit is something I consider to be a dramatic improvement.